Gems/Atom/Feature/Common/Code/Source/CoreLights/SimpleSpotLightFeatureProcessor.cpp

```diff
@@ -141,45 +144,90 @@ namespace AZ
 
             if (m_deviceBufferNeedsUpdate)
             {
+                if (m_goboArrayChanged)
+                {
+                    // collect all gobo textures and assign index for each spot light
+                    AZStd::unordered_map<AZ::Data::Instance<AZ::RPI::Image>, int> gobos;
+
+                    for (int32_t idx = 0; idx < m_lightData.GetDataCount(); idx++)
+                    {
+                        auto& lightData = m_lightData.GetData<0>(idx);
+                        auto image = m_lightData.GetData<1>(idx).m_gobo;
+                        if (image)
+                        {
+                            if (gobos.find(image) == gobos.end())
+                            {
+                                gobos[image] = aznumeric_cast<int>(gobos.size());
+                            }
+                            lightData.m_goboTextureIndex = gobos[image];
+                        }
+                        else
+                        {
+                            lightData.m_goboTextureIndex = MaxGoboTextureCount;
+                        }
+                    }
+                
+                    m_goboTextures.clear();
+                    int32_t currentIdx = 0;
+                    // add gobo textures to the vector. Limit the max gobo texture count.
+                    for (auto& item : gobos)
+                    {
+                        if (currentIdx >= MaxGoboTextureCount)
+                        {
+                            break;
+                        }
```

Comment:
I think an AZ_WarningOnce would be good here if this is the only place devs will be notified that they've exceeded the gobo count